File: java/src/main/java/cn/az/code/year2021/july/PartitionArrayIntoDisjointIntervals.java
package cn.az.code.year2021.july;

/**
 * @author az
 * @since 2021-07-22
 */
public class PartitionArrayIntoDisjointIntervals {
    public int partitionDisjoint(int[] a) {
        int localMax = a[0], partitionIdx = 0, max = localMax;
        for (int i = 1; i < a.length; i++)
            if (localMax > a[i]) {
                localMax = max;
                partitionIdx = i;
            } else
                max = Math.max(max, a[i]);
        return partitionIdx + 1;
    }
}

File: go/2021/july/partition-array-into-disjoint-intervals.go
package july

func partitionDisjoint(nums []int) int {
	localMax, partitionIdx := nums[0], 0
	mx := localMax

	for i := 1; i < len(nums); i++ {
		if localMax > nums[i] {
			localMax = mx
			partitionIdx = i
		} else {
			mx = max(mx, nums[i])
		}
	}
	return partitionIdx + 1
}
